File: dotnet/Common/ReflectionUtil.cs
/*
 * ====================
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 * 
 * Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.     
 * 
 * The contents of this file are subject to the terms of the Common Development 
 * and Distribution License("CDDL") (the "License").  You may not use this file 
 * except in compliance with the License.
 * 
 * You can obtain a copy of the License at 
 * http://IdentityConnectors.dev.java.net/legal/license.txt
 * See the License for the specific language governing permissions and limitations 
 * under the License. 
 * 
 * When distributing the Covered Code, include this CDDL Header Notice in each file
 * and include the License file at identityconnectors/legal/license.txt.
 * If applicable, add the following below this CDDL Header, with the fields 
 * enclosed by brackets [] replaced by your own identifying information: 
 * "Portions Copyrighted [year] [name of copyright owner]"
 * ====================
 */
using System;
using System.Reflection;
using System.Collections.Generic;
using System.Linq;
namespace Org.IdentityConnectors.Common
{
    /// <summary>
    /// Miscellaenous reflection utilities.
    /// </summary>
    public static class ReflectionUtil
    {
        /// <summary>
        /// Gets all the interfaces in the hierarchy
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        public static Type[] GetAllInterfaces(Type type)
        {
            HashSet<Type> temp = new HashSet<Type>();
            GetAllInterfaces2(type, temp);
            return temp.ToArray();
        }

        private static void GetAllInterfaces2(Type type, HashSet<Type> rv)
        {
            if (type != null)
            {
                if (type.IsInterface)
                {
                    rv.Add(type);
                }
                GetAllInterfaces2(type.BaseType, rv);
                foreach (Type inter in type.GetInterfaces())
                {
                    GetAllInterfaces2(inter, rv);
                }
            }
        }

        /// <summary>
        /// Returns the generic type definition of the given type
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        public static Type[] GetTypeErasure(Type[] types)
        {
            Type[] rv = new Type[types.Length];
            for (int i = 0; i < types.Length; i++)
            {
                rv[i] = GetTypeErasure(types[i]);
            }
            return rv;
        }
        /// <summary>
        /// Returns the generic type definition of the given type
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        public static Type GetTypeErasure(Type type)
        {
            if (type.IsGenericType)
            {
                type = type.GetGenericTypeDefinition();
            }
            return type;
        }

        /// <summary>
        /// Returns true iff t1 is a parent type of t2. Unlike
        /// Type.isAssignableFrom, this handles generic types as
        /// well. 
        /// </summary>
        /// <param name="t1"></param>
        /// <param name="t2"></param>
        /// <returns></returns>
        public static bool IsParentTypeOf(Type t1,
                                          Type t2)
        {
            if (t2 == null)
            {
                return t1 == null;
            }
            Type found = FindInHierarchyOf(t1, t2);
            return found != null;
        }

        /// <summary>
        /// Finds t1 in the hierarchy of t2 or null if not found. The
        /// returned value will have generic parameters applied to it.
        /// </summary>
        /// <param name="t1"></param>
        /// <param name="t2"></param>
        /// <returns></returns>
        public static Type FindInHierarchyOf(Type t1, Type t2)
        {
            if (t2 == null)
            {
                return null;
            }
            if (EqualsIgnoreGeneric(t1, t2))
            {
                return t2;
            }
            Type found1 = FindInHierarchyOf(t1, t2.BaseType);
            if (found1 != null)
            {
                return found1;
            }
            foreach (Type inter in t2.GetInterfaces())
            {
                Type found2 = FindInHierarchyOf(t1, inter);
                if (found2 != null)
                {
                    return found2;
                }
            }
            return null;
        }

        private static bool EqualsIgnoreGeneric(Type t1,
                                            Type t2)
        {
            if (t1 == null || t2 == null)
            {
                return t1 == null && t2 == null;
            }
            if (t1.IsGenericType)
            {
                t1 = t1.GetGenericTypeDefinition();
            }
            if (t2.IsGenericType)
            {
                t2 = t2.GetGenericTypeDefinition();
            }
            return t1.Equals(t2);
        }

        public static Type[] GetParameterTypes(MethodInfo method)
        {
            ParameterInfo[] parameters = method.GetParameters();
            Type[] rv = new Type[parameters.Length];
            for (int i = 0; i < parameters.Length; i++)
            {
                rv[i] = parameters[i].ParameterType;
            }
            return rv;
        }
    }
}

File: framework/src/main/java/org/identityconnectors/common/ReflectionUtil.java
/*
 * ====================
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 * 
 * Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.     
 * 
 * The contents of this file are subject to the terms of the Common Development 
 * and Distribution License("CDDL") (the "License").  You may not use this file 
 * except in compliance with the License.
 * 
 * You can obtain a copy of the License at 
 * http://IdentityConnectors.dev.java.net/legal/license.txt
 * See the License for the specific language governing permissions and limitations 
 * under the License. 
 * 
 * When distributing the Covered Code, include this CDDL Header Notice in each file
 * and include the License file at identityconnectors/legal/license.txt.
 * If applicable, add the following below this CDDL Header, with the fields 
 * enclosed by brackets [] replaced by your own identifying information: 
 * "Portions Copyrighted [year] [name of copyright owner]"
 * ====================
 */
package org.identityconnectors.common;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ReflectionUtil {

    /**
     * Never allow this to be instantiated.
     */
    private ReflectionUtil() {
        throw new AssertionError();
    }

    /**
     * Builds a {@link Set} of interfaces from the target class.
     */
    public static Set<Class<?>> getAllInterfaces(final Class<?> target) {
        assert target != null;
        Set<Class<?>> ret = new HashSet<Class<?>>();
        getAllInteralInterfaces(target, ret);
        return ret;
    }

    private static void getAllInteralInterfaces(final Class<?> target,
            final Set<Class<?>> result) {
        // quick exit if target is null..
        if (target != null) {
            // get all the interfaces of the target class..
            for (Class<?> inter : target.getInterfaces()) {
                result.add(inter);
            }
            // get all the interfaces of the super class..
            getAllInteralInterfaces(target.getSuperclass(), result);
        }
    }

    /**
     * Determine if the target class implements the provided interface.
     * 
     * @param target
     *            class to look through for a matching interface.
     * @param clazz
     *            interface class to look for.
     * @return true if a matching interface is found otherwise false.
     */
    public static boolean containsInterface(final Class<?> target,
            final Class<?> clazz) {
        return clazz.isAssignableFrom(target);
    }

    /**
     * Get all interfaces the extends the type provided.
     */
    public static <T> List<Class<? extends T>> getInterfaces(
            final Class<?> target, final Class<T> type) {
        List<Class<? extends T>> ret = new ArrayList<Class<? extends T>>();
        Collection<Class<?>> interfs = getAllInterfaces(target);
        for (Class<?> clazz : interfs) {
            if (containsInterface(clazz, type)) {
                @SuppressWarnings("unchecked")
                Class<? extends T> o = (Class<? extends T>) clazz;
                ret.add(o);
            }
        }
        return ret;
    }
    
    /**
     * Returns true iff the given class overrides equals and hashCode
     * @param clazz The class to check.
     * @return True iff the given class overrides equals and hashCode
     */
    public static boolean overridesEqualsAndHashcode(Class<?> clazz) {
        try {
            Method equals = clazz.getMethod("equals", Object.class);
            if (equals.getDeclaringClass() == Object.class) {
                return false;
            }
            Method hashCode = clazz.getMethod("hashCode");
            if (hashCode.getDeclaringClass() == Object.class) {
                return false;
            }
            return true;
        }
        catch (RuntimeException e) {
            throw e;
        }
        catch (Exception e) {
            //this should never happen
            throw new RuntimeException(e);
        }
    }

    /**
     * Returns the package the class is associated with.
     * 
     * @param clazz
     *            class to inspect for the package.
     * @return package for the class provided.
     * @throws NullPointerException
     *             iff clazz is <code>null</code>.
     */
    public static String getPackage(Class<?> clazz) {
        String name = clazz.getName();
        return name.substring(0, name.lastIndexOf('.'));
    }

    /**
     * Determine the method name for the calling class.
     */
    public static String getMethodName(int depth) {
        // Hack (?) to get the stack trace.
        Throwable dummyException = new Throwable();
        StackTraceElement locations[] = dummyException.getStackTrace();
        // caller will be the depth element
        String method = "unknown";
        if (locations != null && locations.length > depth) {
            StackTraceElement caller = locations[depth];
            method = caller.getMethodName();
        }
        return method;
    }
}
